### PR TITLE
Resolve #29 [Assert] All-in Hamcrest

### LINKED ISSUE #29: [Assert] All-in Hamcrest
Hamcrest mattchers are cool as hell. They are easily maintainable and extensible. So... <sad face>
1. Remove all custom verbose assert performers (such as `org.whaka.asserts.builder.ObjectAssertPerformer<T>`, `org.whaka.asserts.ObjectAssert<T>`, etc.)
2. Remove all excessive methods from `AssertBuilder` and `Assert`
3. Fix javadocs

We're gonna provide a bunch of custom matchers, for things like "equal with ComparisonPerform", or "equal doubles with DoubleMath". `Assert` and `AssertBuilder` will only provide methods to work with matchers. The main purpose of `AssertBuilder` will be to allow to collect multiple assert result, and the main purpose of the `Assert` (Why not use Hamcrest's `MatcherAssert`) is to provide a functionality to easily throw the same type of an exception as the builder. For any new functionality we'll have to implement a single new matcher.

**Note:** custom matchers should be created for any truly unique functionality that old asserts provided. But separate issue should be created for each new matcher. All issues should be linked to this one.

### PR DESCRIPTION
**Pull request by zuzunker**
1. `org.whaka.asserts.builder` package is removed
2. `AssertBuilder` is moved into `org.whaka.asserts`
3. All "assert performers" and "instant asserts" are removed
4. AssertBuilder provides now 3 main methods:
   - `#checkThat(Object, Matcher)`
   - `#checkThat(Object, Matcher, String)`
   - `#checkThat(Object, Matcher, String, Throwable)`
5. Assert provides now 3 main methods:
   - `#assertThat(Object, Matcher)`
   - `#assertThat(Object, Matcher, String)`
   - `#assertThat(Object, Matcher, String, Throwable)`

**Note:** any custom matchers will be implemented under separate issues.

### CODE REVIEW

File: src/main/java/org/whaka/asserts/AssertBuilder.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	
+	/**
+	 * New instance of the AssertError is returned if any assert results are present.
+	 * Otherwise null is returned.
+	 */
+	public AssertError build() {
+		if (assertResults.size() > 0)
+			return new AssertError(assertResults);
+		return null;
+	}
```
Comment (visitorus) [inline src/main/java/org/whaka/asserts/AssertBuilder.java:43]: Maybe it's better to return some `AssertError.EMPTY` instead of `null`?

**Comment by zuzunker** [comment]
> Maybe it's better to return some AssertError.EMPTY instead of null?
1. Should this EMPTY error than be also throwable?
   - What is the point of an "error of the assertion" without any assertion (messages/results)?
   - How then method "#performAssert" should behave in case of empty builder? Should it still throw this EMPTY error?
2. If error shouldn't be fully functional and throwable, then:
   - How?
   - What is the point of it?

It was decided to return `null` in this case, for it seems to be the only proper way to indicate **the lack of error**. Also note, that `AssertError` cannot be empty, this treated as construction exception.

**Comment by visitorus** [comment]
1. To avoid `null` value, it's better to return special case of expected class. Potential NPE will be eliminated. 
2. `#performAssert` will look like: `if(!error.isEmpty()) throw error`.

**Comment by zuzunker** [comment]
> To avoid null value, it's better to return special case of expected class. Potential NPE will be eliminated. 

It sounds good, as everything dogmatic for no reason. But tells nothing, on the subject of how should it work in this particular case.

Does it really pay off to have a system-wide **empty** instance of an error - which doesn't have a point without some proper message in it - exactly the thing I was trying to avoid with exceptions in the error constructor, only to avoid a single **possibility** of an NPE? Especially taking to note the fact that `#build` method is created mostly for consistency matters, and will be actually used less than never.

Trying to fix this "neverhappen" problem will _literally_ introduce more additional complexity than the occasional `null-check` (if method ever will be used, which I doubt).

And still, if we introduce an empty instance we **have** to mind additional usecase of the:

```
throw AssertError.EMPTY;
```

Which is unacceptable.


---

Method might return `Optional<AssertError>`

File: src/main/java/org/whaka/util/Try.java [hunk 2]
```diff
@@ -325,12 +328,12 @@ public boolean isCaught() {
...
+	 * If this try is not successful - {@link Assert#assertThat(Object, Matcher, String)} is called
+	 * with the specified matcher and message. Any errors evoked by the assert are rethrown directly.
 	 */
-	public Try<R> assertFail(Consumer<ThrowableAssert> consumer) {
+	public Try<R> assertFail(Matcher<? super Exception> matcher, String message) {
 		if (!isSuccess())
-			consumer.accept(Assert.assertThrowable(getCause()));
+			Assert.assertThat(getCause(), matcher, message);
 		return this;
 	}
```
Comment (visitorus) [inline src/main/java/org/whaka/util/Try.java:350]: Add `assertFail(Consumer<ThrowableAssert> consumer)` (without message)

**Comment by zuzunker** [comment]
> Add assertFail(Consumer<ThrowableAssert> consumer) (without message)

Done

**Comment by visitorus** [comment]
"Optional" looks better.

P.S.
You don't have to write tons of sarcasm in comment. "null" from the "build" method looks weird.

**Comment by zuzunker** [comment]
> "Optional" looks better.

Fixed

> You don't have to write tons of sarcasm in comment. "null" from the "build" method looks weird.

Noted. Actually, only part about "dogmatic" was a sarcasm =)
Otherwise I do always appreciate a discussion of a proposed code, but just am doing it weird.
I do agree about `null` in `#build` method, but have forgotten about optional at all.

**Pull request merged**